<div>
    <img src = "https://www.rd.com/wp-content/uploads/2019/03/shutterstock_1113632312-scaled.jpg" width="800">
</div>

# Mobile App for Lottery Addiction

Gambling can be very fun for many people. For some, however, it can become a habit, which then leads to addiction. Addiction to gambling can cause desperate behaviors, such as spending savings and loans, accumlating debt, and even theft.

A medical institute wants to try and combat this problem through the used of a dedicated mobile app to help lottery addicts better estimate their chances of winning. The institute already has engineers to build the app, but they want us to calculate some probabilities.

For the first version of the app, they want us to focus on the 6/49 lottery and build functions that enable users to answer questions like:

* What is the probability of winning the big prize with a single ticket?
* What is the probability of winning the big prize if we play 40 different tickets (or any other number)?
* What is the probability of having at least five (or four, or three, or two) winning numbers on a single ticket?

Because we're dealing with probabilities and combinations, it will be helpful to have some functions that will do the math for us:


In [1]:
def factorial(n):
    final = 1
    for i in range(n,0,-1):
        final = final * i
    return final

def permutation(n,k):
    return (factorial(n) / factorial(n-k))

def combinations(n,k):
    return factorial(n) / (factorial(k)*factorial(n-k))

In the 6/49 lottery, players pick six different numbers from the set of numbers from 1 to 49. If the same six numbers are drawn, that player wins the big prize. The player must have all six numbers matching the numbers drawn in order to win. If only a single number is different, he doesn't win.

For the first version of the app, we want the players to be able to simulate their probability of winning the big prize with the various numbers they like to play on a single ticket (one ticket = six chosen numbers). Our first task then is to build a function that calculates the probability of winning the big prize for any given ticket.

Before moving forward, the engineering team at the medical institute let us know a few things:

* In the app, the user inputs six different numbers from 1 to 49.
* Under the hood, the six numbers will be in the form of a Python list.
* The engineering team wants the probability to be displayed in a friendly way - in a way that people without any knowledge or probability training could understand.

## Single-Ticket Probability
Below is our function to calculate the probability of winning the big prize on a single ticket, followed by some test inputs.

In [2]:
def one_ticket(nums):
    total_outcomes = combinations(49,len(nums))
    prob_single = 1/total_outcomes
    pct = prob_single * 100
    print('The probability of winning with the numbers {} is: {:.7f}%. In other words, you have 1 in {:,} chances to win.'.format(nums, pct, int(total_outcomes)))

In [3]:
test1 = [2,5,9,34,23,22]
one_ticket(test1)

The probability of winning with the numbers [2, 5, 9, 34, 23, 22] is: 0.0000072%. In other words, you have 1 in 13,983,816 chances to win.


In [4]:
test2 = [4,8,15,16,23,42]
one_ticket(test2)

The probability of winning with the numbers [4, 8, 15, 16, 23, 42] is: 0.0000072%. In other words, you have 1 in 13,983,816 chances to win.


## A Single Ticket Compared To Historical Lottery Data

For the first version of the app, users will also be able to compare their chosen numbers to the entire history of the 6/49 Lottery to see if their numbers have ever been picked.

In [5]:
data = pd.read_csv('649.csv')
print(data.shape)
data.head(3)

(3665, 11)


PRODUCT  DRAW NUMBER  SEQUENCE NUMBER  DRAW DATE  NUMBER DRAWN 1  \
0      649            1                0  6/12/1982               3   
1      649            2                0  6/19/1982               8   
2      649            3                0  6/26/1982               1   

   NUMBER DRAWN 2  NUMBER DRAWN 3  NUMBER DRAWN 4  NUMBER DRAWN 5  \
0              11              12              14              41   
1              33              36              37              39   
2               6              23              24              27   

   NUMBER DRAWN 6  BONUS NUMBER  
0              43            13  
1              41             9  
2              39            34

In [6]:
data.tail(3)

PRODUCT  DRAW NUMBER  SEQUENCE NUMBER  DRAW DATE  NUMBER DRAWN 1  \
3662      649         3589                0  6/13/2018               6   
3663      649         3590                0  6/16/2018               2   
3664      649         3591                0  6/20/2018              14   

      NUMBER DRAWN 2  NUMBER DRAWN 3  NUMBER DRAWN 4  NUMBER DRAWN 5  \
3662              22              24              31              32   
3663              15              21              31              38   
3664              24              31              35              37   

      NUMBER DRAWN 6  BONUS NUMBER  
3662              34            16  
3663              49             8  
3664              48            17

We can see that we have data for 3,665 lottery draws, along with the date and the order in which they were drawn. Now we're going to create a function that will enable users to compare their ticket against the historical data and determine whether they would have ever won by now. The engineering team wants us to print:
* The number of times the combination selected occured in the Canada data set.
* the probability of winning the big prize in the next drawing with that combination.

In [7]:
def extract_numbers(row):
    row = row[4:10]
    row = set(row.values)
    return row

In [8]:
winners = data.apply(extract_numbers, axis=1)

In [9]:
def check_historical_occurence(user_nums,winners):
    numbers = set(user_nums)
    occurrence = numbers == winners
    n_occurrence = occurrence.sum()
    
    if n_occurrence == 0:
        print('The numbers {} have never been drawn.'.format(user_nums))
        print('This does not make it more likely that they will be drawn soon.')
        print('Your chances to win the big prize in the next drawing using {} are 0.0000072%.'.format(user_nums))
        print('In other words, you have a 1 in 13,983,816 chances to win.')
    else:
        if n_occurrence == 1:
            print('The numbers {} have been drawn {} time in the past.'.format(user_nums, n_occurrence))
        else:
            print('The numbers {} have been drawn {} times in the past.'.format(user_nums, n_occurrence))
        print('Your chances to win the big prize in the next drawing using {} are 0.0000072%.'.format(user_nums))
        print('In other words, you have a 1 in 13,983,816 chances to win.')

In [10]:
test4 = [4,8,15,16,23,42]
check_historical_occurence(test4,winners)

The numbers [4, 8, 15, 16, 23, 42] have never been drawn.
This does not make it more likely that they will be drawn soon.
Your chances to win the big prize in the next drawing using [4, 8, 15, 16, 23, 42] are 0.0000072%.
In other words, you have a 1 in 13,983,816 chances to win.


## Multi-Ticket Probability

People addicted to the lottery are most likely going to play more than one ticket on a single drawing. The purpose of this app is to help them better understand their chances of winning. We're going to write a new function that will allow the user to calculate the chances of winning for any number of different tickets. Again, the engineering team has given us some helpful information:

* The user will input the *number* of different tickets they'd like to play (without inputting the specific combination of numbers).
* Our function will see that as a number between 1 and 13,983,816 (the maximum possible number of unique tickets).
* The function should print information about the probability of winning the big prize depending on the number of different tickets played.

In [11]:
total_outcomes = combinations(49,6)
def multi_ticket_probability(num_tickets):
    multi_prob = (num_tickets / total_outcomes) * 100
    if num_tickets == 1:
        print('The probability of winning by playing 1 ticket is: {:.7f}%.'.format(multi_prob))
    else:
        print('The probability of winning by playing {} tickets is: {:.7f}%.'.format(num_tickets, multi_prob))
    combos = total_outcomes / num_tickets
    print('In other words, you have 1 in {:,} chances to win.'.format(int(combos)))

In [12]:
test5 = [1, 10, 100, 10000, 1000000, 6991908, 13983816]
for i in test5:
    multi_ticket_probability(i)
    print('------------------')

The probability of winning by playing 1 ticket is: 0.0000072%.
In other words, you have 1 in 13,983,816 chances to win.
------------------
The probability of winning by playing 10 tickets is: 0.0000715%.
In other words, you have 1 in 1,398,381 chances to win.
------------------
The probability of winning by playing 100 tickets is: 0.0007151%.
In other words, you have 1 in 139,838 chances to win.
------------------
The probability of winning by playing 10000 tickets is: 0.0715112%.
In other words, you have 1 in 1,398 chances to win.
------------------
The probability of winning by playing 1000000 tickets is: 7.1511238%.
In other words, you have 1 in 13 chances to win.
------------------
The probability of winning by playing 6991908 tickets is: 50.0000000%.
In other words, you have 1 in 2 chances to win.
------------------
The probability of winning by playing 13983816 tickets is: 100.0000000%.
In other words, you have 1 in 1 chances to win.
------------------


## N-digit Probability

We're going to create one more function that will allow the user to calculate the probabilities for two, three, four, or five winning numbers. In most 6/49 lotteries, players can win a smaller prize if they match two, three, four, or five of the numbers drawn. The users might be interested in known the probabilities for certain amounts of numbers. The engineering team has again given us the following information:
* In the app, the user will input:
    * Six different numbers from 1 to 49, and
    * an integer between 2 and 5 that represents the number of winning numbers expected.
* The function will print the probability of having the inputted number of winning numbers.

In [13]:
def probability_less_6(num):
    combos = combinations(6,num)
    possible_wins = combinations(43, 6-num)
    total_wins = combos * possible_wins
    probability = total_wins / total_outcomes * 100
    combos_total = round(total_outcomes / total_wins)
    print('The chance of winning by matching {} numbers is {:.7f}%'.format(num,probability))
    print('In other words, you have a 1 in {} chance of winning.'.format(combos_total))

In [14]:
test6 = [2,3,4,5]
for i in test6:
    probability_less_6(i)
    print('------------------')

The chance of winning by matching 2 numbers is 13.2378029%
In other words, you have a 1 in 8 chance of winning.
------------------
The chance of winning by matching 3 numbers is 1.7650404%
In other words, you have a 1 in 57 chance of winning.
------------------
The chance of winning by matching 4 numbers is 0.0968620%
In other words, you have a 1 in 1032 chance of winning.
------------------
The chance of winning by matching 5 numbers is 0.0018450%
In other words, you have a 1 in 54201 chance of winning.
------------------


In the first version of our app, we have programmed four functions:

* `one_ticket_probability()` — calculates the probability of winning the big prize with a single ticket
* `check_historical_occurrence()` — checks whether a certain combination has occurred in the Canada lottery data set
* `multi_ticket_probability()` — calculates the probability for any number of of tickets between 1 and 13,983,816
* `probability_less_6()` — calculates the probability of having two, three, four or five winning numbers

## Probability of Having Less Winning Numbers - At Least N

Finally, let's consider the case when the user is interested in knowing the probability of having *at least* two, three, four, or five winning numbers. This function will be similar to our previous function, except that the number of successful outcomes for having at least N winning numbers will be the sum of the numbers of successful outcomes for having exactly N winning numbers, N+1, etc., up to and including 6.

In [15]:
def probability_at_least_num(num):
    successes = 0
    for n in range(num,7):
        comb = combinations(6,n)
        lottery_outcomes = combinations(43, 6-n)
        successes += comb * lottery_outcomes
    probability = successes / total_outcomes * 100
    combos_total = round(total_outcomes / successes)
    print('The chance of winning by having at least {} numbers is {:.7f}%'.format(num,probability))
    print('In other words, you have a 1 in {} chance of winning.'.format(combos_total))

In [16]:
test7 = [2,3,4,5]
for i in test7:
    probability_at_least_num(i)
    print('------------------')

The chance of winning by having at least 2 numbers is 15.1015574%
In other words, you have a 1 in 7 chance of winning.
------------------
The chance of winning by having at least 3 numbers is 1.8637545%
In other words, you have a 1 in 54 chance of winning.
------------------
The chance of winning by having at least 4 numbers is 0.0987141%
In other words, you have a 1 in 1013 chance of winning.
------------------
The chance of winning by having at least 5 numbers is 0.0018521%
In other words, you have a 1 in 53992 chance of winning.
------------------
